### PR TITLE
feat: graceful shutdown

### PR DESCRIPTION
**Pull request by sammy1991106**
## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues
- [ ] Tag type of change

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -116,25 +119,29 @@ impl<'a> App<'a> {
...
-        let mut state = self.state;
+        let event_sub_driver = self.event_sub_driver;
 
         let event_sub_handle = tokio::spawn(async move { client.run().await });
         let event_processor_handle = tokio::spawn(async move { processor.run().await });
         let broadcaster_handler = tokio::spawn(async move { broadcaster.run().await });
 
-        self.state_updater.run(&mut state).await;
+        tokio::spawn(async move {
+            signal::ctrl_c().await.expect("failed to capture signal");
```
Comment (cgorenflo) [inline src/lib.rs:133]: good start, but that's not the only way the program can get terminated. Use signal_hook crate instead and catch SIGINT & SIGTERM?

Comment (cgorenflo) [reply src/lib.rs:133]: https://dev.to/talzvon/handling-unix-kill-signals-in-rust-55g6

File: src/lib.rs [hunk 2]
```diff
@@ -116,25 +119,29 @@ impl<'a> App<'a> {
...
-        self.event_sub_driver.close().map_err(Error::new)?;
-        state.flush().map_err(Error::new)?;
+        self.state_updater.run(self.state).await.map_err(Error::new)?;
 
-        let _ = event_sub_handle.await;
-        let _ = event_processor_handle.await;
-        let _ = broadcaster_handler.await;
+        event_sub_handle.await.map_err(Error::new)?.map_err(Error::new)?;
+        event_processor_handle.await.map_err(Error::new)?.map_err(Error::new)?;
+        broadcaster_handler.await.map_err(Error::new)?.map_err(Error::new)?;
```
Comment (cgorenflo) [inline src/lib.rs:144]: what's the double mapping for?

File: src/queue/queued_broadcaster.rs [hunk 3]
```diff
@@ -121,6 +121,8 @@ where
             }
         }
 
+        broadcast_all(&mut queue, &mut broadcaster).await?;
```
Comment (cgorenflo) [inline src/queue/queued_broadcaster.rs:124]: just checking, this is a last "try to broadcast all remaining messages before termination" call?

Comment (sammy1991106) [reply src/queue/queued_broadcaster.rs:124]: That's absolutely right.

File: src/lib.rs (hunk 1 above)
Comment (sammy1991106) [reply src/lib.rs:133]: I thought that ".ctrl_c" handles both SIGINT and SIGTERM but just found out I was wrong. Updated.

**Review by cgorenflo — APPROVED** [review]

**Review by haiyizxx — APPROVED** [review]

**Pull request merged**